### ISSUE TITLE
YOLO2 takes too much memory, cannot train on GPU

### ISSUE THREAD

**Issue by saudet** [deeplearning4j/deeplearning4j#4784]
#### Issue Description

When training with YOLO2 on GPU, device memory fills up and then it gives up with the log below on that test in deeplearning4j-zoo. On CPU it takes 34 GB of RAM. By comparison, TinyYOLO takes about 6 GB of GPU RAM and runs with 6 GB on CPU. AFAIK, there are no GPUs with 34 GB of RAM...

```
Running org.deeplearning4j.zoo.TestInstantiation
o.d.z.TestInstantiation - Testing training on zoo model YOLO2
o.n.l.f.Nd4jBackend - Loaded [JCublasBackend] backend
o.n.n.NativeOpsHolder - Number of threads used for NativeOps: 32
o.n.n.Nd4jBlas - Number of threads used for BLAS: 0
o.n.l.a.o.e.DefaultOpExecutioner - Backend used: [CUDA]; OS: [Linux]
o.n.l.a.o.e.DefaultOpExecutioner - Cores: [6]; Memory: [12.2GB];
o.n.l.a.o.e.DefaultOpExecutioner - Blas vendor: [CUBLAS]
o.n.l.j.o.e.CudaExecutioner - Device opName: [Tesla K80]; CC: [3.7]; Total/free memory: [11995578368]
o.d.n.g.ComputationGraph - Starting ComputationGraph with WorkspaceModes set to [training: SEPARATE; inference: SEPARATE]
Tests run: 3, Failures: 0, Errors: 1, Skipped: 2, Time elapsed: 78.054 sec <<< FAILURE! - in org.deeplearning4j.zoo.TestInstantiation
testMultipleCnnTraining(org.deeplearning4j.zoo.TestInstantiation)  Time elapsed: 77.741 sec  <<< ERROR!
java.lang.RuntimeException: Failed to allocate 94633984 bytes from DEVICE [0] memory
    at org.deeplearning4j.zoo.TestInstantiation.testMultipleCnnTraining(TestInstantiation.java:65)
```

After 3 calls to `fit()` on CPU, we get the following output from  `Nd4j.getWorkspaceManager().printAllocationStatisticsForCurrentThread()`:
```
o.n.l.m.p.BasicWorkspaceManager - Workspace statistics: ---------------------------------
o.n.l.m.p.BasicWorkspaceManager - Number of workspaces in current thread: 3
o.n.l.m.p.BasicWorkspaceManager - Workspace: LOOP_CACHE
o.n.l.m.p.BasicWorkspaceManager - Allocated amount: 0 bytes
o.n.l.m.p.BasicWorkspaceManager - External (spilled) amount: 0 bytes
o.n.l.m.p.BasicWorkspaceManager - External (pinned) amount: 0 bytes

o.n.l.m.p.BasicWorkspaceManager - Workspace: LOOP_EXTERNAL
o.n.l.m.p.BasicWorkspaceManager - Allocated amount: 34248563558 bytes
o.n.l.m.p.BasicWorkspaceManager - External (spilled) amount: 0 bytes
o.n.l.m.p.BasicWorkspaceManager - External (pinned) amount: 0 bytes

o.n.l.m.p.BasicWorkspaceManager - Workspace: LOOP_FF
o.n.l.m.p.BasicWorkspaceManager - Allocated amount: 4542431232 bytes
o.n.l.m.p.BasicWorkspaceManager - External (spilled) amount: 0 bytes
o.n.l.m.p.BasicWorkspaceManager - External (pinned) amount: 0 bytes
```

According to @raver119, this indicates that most of the memory is being hogged by activations...

#### Version Information

* Deeplearning4j 0.9.2-SNAPSHOT
* Ubuntu 16.04.3 LTS
* CUDA 8.0.61
* NVIDIA driver 384.66

/cc @DaGump

**Comment by AlexDBlack** [comment]
Looking at this for input size 608x608:

Parameters length: 50,983,561
Activations length, 1 example: 110,651,554

Parameters length (bytes): 194.49 M
Activations length (bytes): 422.10 M

Minimum memory requirements, inference, FP32 (returning *all* activations):
Fixed: 194.49M
Per example: 422.10 MB

Approximate minimum memory requirements, training, FP32 (assuming current design)
Fixed: params + parameter gradients + updater state (Adam) = 4 * Params = 778 MB
Per example: Activations + activation gradients = 2 * Per example = 844 MB
Plus working memory overheads, async prefetch, labels, etc...

12GB GPU: Max minibatch size of 13 examples (best case - upper bound)

Note that in both forward and backward cases, there are things we can do to reduce the memory requirements - in some cases, at the expense of performance.

**Comment by AlexDBlack** [comment]
So, the next question: how much memory are we seeing in practice?

----
Forward pass (4 times - keeping all activations in memory, with workspaces enabled):
Total JavaCPP bytes: (Pointer.totalBytes()): 1.32 G
LOOP_FF workspace: Allocated amount: 227121715 bytes (227MB)
Note that the model does *not* have updater or gradients or anything (as expected).
This memory use isn't unreasonable, and is actually less than I expected given the calculations above. Why is it less? Not sure. Maybe some of the layers are either no-op or in-place (hence I'm double counting).

-----

Backward pass - (4x fit operations)

Minibatch size: 1, 2, 4
LOOP_EXTERNAL: 304MB, 534MB, 996MB
LOOP_FF: 255MB, 567MB, 1135MB
Max Pointer.totalBytes(): 906MB, 962MB, 1.04GB
Max Pointer.physicalBytes(): 4.47GB, 7.22GB, 12.53GB

I get OOM at minibatch size 5 with default XMX settings (7GB) on a 32GB system

Anyway, in conclusion: we are a ways off the theoretical minimum here in practice (without doing anything fancy - of which we have a few options), by a factor of about 3x for each example in minibatch. So yes, there's definitely scope for improvement here. But it's not going to be a simple solution/fix to do that.

Test code: https://gist.github.com/AlexDBlack/702ebc57565475564118b53dee7a1c87

**Comment by saudet** [comment]
/cc @jpatanooga 

@raver119 Do you see any possible "quick fixes"?

**Comment by AlexDBlack** [comment]
FYI I've posted some thoughts on basecamp regarding approaches to reduce memory requirements for training.

**Comment by treo** [comment]
I've found that most of the updaters can use quite a bit of memory temporarily. The Adam updater for example uses 3x Params for transient calculations.

**Comment by AlexDBlack** [comment]
@treo hm, that's an interesting and very useful observation, thanks.
Note that we combine updaters across layers to maximize performance... in the worst case (for memory, best case for performance), the entire network params can be handled by *one* Adam/Nesterov/etc object. So one simple thing here would be to set a maximum size on the updater blocks...
https://github.com/deeplearning4j/deeplearning4j/blob/4ed60e938c39a2732586290e526b69d9d821c274/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/updater/BaseMultiLayerUpdater.java#L59-L65

**Comment by AlexDBlack** [comment]
Should be resolved now, as of: https://github.com/deeplearning4j/deeplearning4j/pull/4900
Memory use is now considerably lower than 0.9.1 and 1.0.0-alpha, especially for ComputationGraph.
